src/OpportunisticResourceEstimator.cpp

```diff
@@ -0,0 +1,122 @@
+#include "OpportunisticResourceEstimator.hpp"
+#include "CommandRunner.hpp"
+#include "Helpers.hpp"
+
+#include <google/protobuf/text_format.h>
+
+#include <process/defer.hpp>
+#include <process/dispatch.hpp>
+
+namespace criteo {
+namespace mesos {
+
+using namespace process;
+using namespace mesos;
+
+using std::string;
+
+using ::mesos::modules::Module;
+using ::mesos::Resource;
+using ::mesos::Resources;
+using ::mesos::slave::ResourceEstimator;
+using ::mesos::ResourceUsage;
+
+using process::Future;
+using process::Failure;
+using process::terminate;
+
+class OpportunisticResourceEstimatorProcess
+    : public process::Process<OpportunisticResourceEstimatorProcess> {
+ public:
+  OpportunisticResourceEstimatorProcess(
+      const lambda::function<Future<ResourceUsage>()>& _usage,
+      const Option<Command>& oversubscribableCommand, bool isDebugMode);
+
+  virtual process::Future<Resources> oversubscribable();
+  virtual process::Future<Resources> _oversubscribable(
+      const ResourceUsage& usage);
+
+ protected:
+  const lambda::function<Future<ResourceUsage>()> usage;
+
+ private:
+  Option<Command> m_oversubscribableCommand;
+  bool m_isDebugMode;
+};
+
+OpportunisticResourceEstimatorProcess::OpportunisticResourceEstimatorProcess(
+    const lambda::function<process::Future<ResourceUsage>()>& _usage,
+    const Option<Command>& oversubscribableCommand, bool isDebugMode)
+    : usage(_usage),
+      m_oversubscribableCommand(oversubscribableCommand),
+      m_isDebugMode(isDebugMode) {}
+
+Future<Resources> OpportunisticResourceEstimatorProcess::_oversubscribable(
+    const ResourceUsage& usage) {
+  // Mocking a resources to have valid return
+  Try<Resources> _resources = ::mesos::Resources::parse(
+      "[{\"name\" : \"cpus\", \"type\":\"SCALAR\", \"scalar\" : {\"value\" : "
+      "\"0\"}, \"role\" : \"*\"}]");
+  Resources resources = _resources.get();
```

Comment:
do we need to return a valid resource if to set?
Shouldn't we return something like a Failure (the "null" case of Try class)

Reply:
@treefell could you answer my comment here?

Reply:
Actually try to return **NULL** and **nullptr** it does not compile while i do that so i went for resources with cpu at 0 as values. I will try to find another way.

Reply:
Try<Resources> means you can return a "success" with resources or a "failure" with an error message. Try returning a failure